onnxruntime/core/providers/cpu/llm/rotary_embedding_helper.h

```diff
@@ -0,0 +1,196 @@
+    }
+    const auto& sin_cache_dims = sin_cache->Shape().GetDims();
+    if (sin_cache_dims.size() != 3) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'sin_cache' is expected to have 3 dimensions, got ",
+                             sin_cache_dims.size());
+    }
+    if (cos_cache_dims[0] != sin_cache_dims[0] || cos_cache_dims[1] != sin_cache_dims[1] || cos_cache_dims[2] != sin_cache_dims[2]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 'cos_cache' and 'sin_cache' are expected to have ",
+                             "the same shape");
+    }
+
+    max_sequence_length = static_cast<int>(cos_cache_dims[1]);
+    head_size = rotary_embedding_dim == 0 ? static_cast<int>(cos_cache_dims[2]) * 2
+                                          : static_cast<int>(hidden_size / num_heads);
+    if (rotary_embedding_dim > 0 && rotary_embedding_dim > head_size) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "rotary_embedding_dim must be less than or equal to ",
+                             "head_size");
+    }
+    // Check cos_cache input shapes
+    if (max_sequence_length != static_cast<int>(cos_cache_dims[1])) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'cos_cache' dimension 1 should be same as ",
+                             "max_sequence_length, got ", cos_cache_dims[0]);
+    }
+    if ((head_size / 2) != static_cast<int>(cos_cache_dims[2]) && (rotary_embedding_dim > 0 && (rotary_embedding_dim / 2) != static_cast<int>(cos_cache_dims[2]))) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'cos_cache' dimension 2 should be same as ",
+                             "head_size / 2 or rotary_embedding_dim / 2, got ", cos_cache_dims[1]);
+    }
+  } else {
+    // Check cos_cache and sin_cache
+    const auto& cos_cache_dims = cos_cache->Shape().GetDims();
+    if (cos_cache_dims.size() != 2) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'cos_cache' is expected to have 2 dimensions, got ",
+                             cos_cache_dims.size());
+    }
+    const auto& sin_cache_dims = sin_cache->Shape().GetDims();
+    if (sin_cache_dims.size() != 2) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'sin_cache' is expected to have 2 dimensions, got ",
+                             sin_cache_dims.size());
+    }
+    if (cos_cache_dims[0] != sin_cache_dims[0] || cos_cache_dims[1] != sin_cache_dims[1]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 'cos_cache' and 'sin_cache' are expected to have ",
+                             "the same shape");
+    }
+    // Check position_ids
+    const auto& position_ids_dims = position_ids->Shape().GetDims();
+    if (!onnxruntime::IsScalarOr1ElementVector(position_ids) && position_ids_dims.size() != 2) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'position_ids' is expected to have 0, 1, or 2 ",
```

Comment:
Do we allow scalar (0 dimension) for position_ids? If so, update line 38

Reply:
Done